### PR TITLE
Support ASCII parameter binding

### LINKED ISSUE #4263: Support true VARCHAR binding to prevent implicit conversions?
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes
| RFC         | no
| BC Break    | no

#### Summary

PHP 7.2 added support for specifying different bind types for unicode and non-unicode string data: https://wiki.php.net/rfc/extended-string-types-for-pdo

This is a major improvement over the default (PDO::PARAM_STR) which binds strings as unicode data. In the past, the inability for PHP to handle this has led to performance problems in certain cases. For example if we have something like:

```php
$statement = $connection->prepare(
   'SELECT * FROM ... WHERE myVarcharColumn = :value' -- will bind as N'...'
);

$statement->bindValue(':value', $myVarcharValue);
$statement->execute();
```

If there's an index on `myVarcharColumn` SQL won't be able to use the index effectively since the binding will be an NVARCHAR string and force an [implicit conversation](http://blog.sqlgrease.com/are-varchar-or-char-hurting-your-performance-due-to-implicit-conversions/).

Does the team have a plan to support this binding in the future? We actually implemented this internally in a wrapper we have DBAL. I'm happy to open a PR if the team supports this feature:

```php
use Doctrine\DBAL\Platforms\AbstractPlatform;
use Doctrine\DBAL\Types\StringType;
use PDO;

class VarcharType extends StringType
{
    public const NAME = 'varchar';

    public function getSQLDeclaration(array $field, AbstractPlatform $platform)
    {
        $length = $field['length'] ?? $platform->getVarcharDefaultLength();
        return 'VARCHAR(' . $length . ')';
    }

    public function getBindingType()
    {
        return PDO::PARAM_STR | PDO::PARAM_STR_CHAR;
    }

    public function getName()
    {
        return static::NAME;
    }
}
```

### PR DESCRIPTION
**Pull request by gjdanis**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no
| Fixed issues | #4263 

#### Summary

Implements the changes proposed in #4263.

### CODE REVIEW

File: .gitignore
Comment (greg0ire) [inline .gitignore:2]: Please drop that commit. I'd rather not open the door to adding rules for every editor out there. Instead, developers should globally ignore their editor directory if any once in their life and be done with it.

Comment (gjdanis) [reply .gitignore:2]: Thanks, @greg0ire I'll do that.

Unrelated to this, is there a way to run the integration tests locally? The test I've added does return the expected `varchar` on a database I have with this version of the driver, so I'm not sure what's going wrong here.

Comment (greg0ire) [reply .gitignore:2]: There is no easy, documented way to run all tests for all databases if that's what you want, but you can still look at `.travis.yaml`, pick and job and runs what it runs.

Comment (gjdanis) [reply .gitignore]: Thanks @greg0ire - I think I've fixed the issue (misunderstood what `fetchOne` returns).

Maybe in a subsequent PR we could add a docker-compose configuration to make it easier to run integration tests locally?

Comment (greg0ire) [reply .gitignore]: > Maybe in a subsequent PR we could add a docker-compose configuration to make it easier to run integration tests locally?

IIRC, that kind of proposal is rejected on grounds that it it does not also run in the CI, it's going to break without us noticing. But I think some documentation on this, maybe in CONTRIBUTING.md would be welcome

**Comment by greg0ire** [comment]
You should use `git rebase -i origin/3.0.x` and drop any commit related to `.idea`, as asked in https://github.com/doctrine/dbal/pull/4274#discussion_r488220780

Also, please associate your email address with your Github account, or change the
email in your commits to an address already associated with it. If you do not
want to expose your personal email address, you may use
`$userName@users.noreply.github.com`, that way we can still reach you through Github.
Additionally, you will get credit for this contribution on your Github profile.

File: .gitignore
Comment (greg0ire) [inline .gitignore:11]: Why does this file appear in the diff?

File: ci/travis/install-mssql-pdo_sqlsrv.sh
Comment (greg0ire) [inline ci/travis/install-mssql-pdo_sqlsrv.sh:7]: This looks like a good change, but does it related to your changes? Shouldn't it be committed separately?

File: src/ParameterType.php
Comment (greg0ire) [inline src/ParameterType.php:45]:
```suggestion
     * Represents an ascii string data type
```

File: tests/Types/AsciiStringTest.php
Comment (greg0ire) [inline tests/Types/AsciiStringTest.php:44]:
```suggestion
     * @return array<int, array{string, array<string, mixed>}>
```

File: tests/Functional/Types/StringBindingTest.php
Comment (greg0ire) [inline tests/Functional/Types/StringBindingTest.php:30]: This could also be implemented for MySQL, right?

>  MySQL and Microsoft SQL Server use it to store Unicode data. 

https://wiki.php.net/rfc/extended-string-types-for-pdo

**Review by greg0ire — COMMENTED** [review]
docs/en/reference/types.rst should be updated, shouldn't it?

File: tests/Functional/Types/StringBindingTest.php
Comment (gjdanis) [reply tests/Functional/Types/StringBindingTest.php:30]: Yes possibly - do you have a good way to test this? The method described [here](https://stackoverflow.com/questions/1223211/get-the-type-of-a-variable-in-mysql) doesn't work unfortunately.

Comment (greg0ire) [reply tests/Functional/Types/StringBindingTest.php:30]: I don't, sorry. I just meant to clarify why you didn't do it in this PR (which is fine)

Comment (gjdanis) [reply tests/Functional/Types/StringBindingTest.php:30]: Ah perfect - I can open a subsequent issue on this.

File: src/Types/AsciiStringType.php
Comment (morozov) [inline src/Types/AsciiStringType.php:22]: The SQL should be generated by the platform, not the type. Otherwise, it's not extensible.

I propose the following:
1. Introduce `AbstractPlatform::getAsciiStringTypeDeclarationSQL()` that would call `AbstractPlatform::getVarcharTypeDeclarationSQL()`.
2. Override the implementation in `SQLServer2012Platform`.

File: tests/Functional/Types/StringBindingTest.php
Comment (morozov) [inline tests/Functional/Types/StringBindingTest.php:29]: You can check for `AbstractSQLServerDriver` instead.

File: tests/Types/AsciiStringTest.php
Comment (morozov) [inline tests/Types/AsciiStringTest.php:23]: Instead of mocking the platform, I believe the test should use the SQLServer2012Platform that will actually implement the generation of the desired SQL.

File: tests/Functional/Types/StringBindingTest.php
Comment (morozov) [inline tests/Functional/Types/StringBindingTest.php:5]: This test is no longer about the data type, it's about the `ParameterType`. Might make sense to rename to `\Doctrine\DBAL\Tests\Functional\ParameterTypes\AsciiStringTest.php`.

File: tests/Functional/Types/StringBindingTest.php
Comment (morozov) [inline tests/Functional/Types/StringBindingTest.php:11]: There's no longer string binding. It's an `AsciiStringTest`.

**Review by morozov — COMMENTED** [review]
Great work so far!

Besides a couple of comments below, it'd be nice to have another functional test at the driver level that would demonstrate the newly introduced type will work on all platforms, not only on the ones that actually support the new binding type.

See [`BinaryTest`](https://github.com/doctrine/dbal/blob/f2f611779a7b044cf06db6915d5d9f6dd958a3c2/tests/Functional/Types/BinaryTest.php) for the reference.

File: tests/Types/AsciiStringTest.php
Comment (gjdanis) [reply tests/Types/AsciiStringTest.php:23]: I think we can actually test the SQL declaration in `SQLServerPlatformTest`. This test then becomes somewhat not needed.

Added this in the latest push if you want to take a look.

Comment (gjdanis) [reply tests/Types/AsciiStringTest.php]: Unrelated to this do you have any idea why this pipeline step is failing? https://ci.appveyor.com/project/doctrine/dbal/builds/35249676/job/8ytj6a8vv8a8lcv0

It seems like the tests are passing but failing to download some code coverage analysis. Do you have any visibility into whether these failures are related to my changes?

Comment (morozov) [reply tests/Types/AsciiStringTest.php]: It's a known issue, unrelated to your changes.

File: src/Platforms/AbstractPlatform.php
Comment (morozov) [inline src/Platforms/AbstractPlatform.php:196]: Let's keep the naming consistent. It should be `getAsciiStringTypeDeclarationSQL()` (https://github.com/doctrine/dbal/pull/4274#discussion_r489842653).

File: src/Types/AsciiType.php
Comment (morozov) [inline src/Types/AsciiType.php:10]: Please name `AsciiStringType`.

File: tests/Functional/Types/AsciiTest.php
Comment (morozov) [inline tests/Functional/Types/AsciiTest.php:11]: Please name `AsciiStringTypeTest` according to the class being tested.

File: tests/Platforms/SQLServerPlatformTest.php
Comment (morozov) [inline tests/Platforms/SQLServerPlatformTest.php:55]: This could be moved to the `AbstractPlatformTest` and cover all platforms. Only the data provider will have to be overridden here (see other platform tests).

File: tests/Types/AsciiTest.php
Comment (morozov) [inline tests/Types/AsciiTest.php:12]: Please name `AsciiStringTypeTest` according to the class being tested.

Comment (gjdanis) [reply tests/Types/AsciiTest.php:12]: @morozov should be name this new type `AsciiType` or `AsciiStringType`? I tend to prefer the former but let me know either way.

Comment (morozov) [reply tests/Types/AsciiTest.php:12]: I'd prefer `AsciiStringType` and `'ascii_string'`:
1. They extend the String type.
2. Technically, there may be other ASCII-based types (e.g. ASCII Text).

File: src/ParameterType.php
Comment (morozov) [inline src/ParameterType.php:47]: No, this should stay `ASCII`. This is a binding type that could be reused by different data types (e.g. `'ascii_text'`). The `'ascii_string'` is a data type.

File: src/Types/Type.php
Comment (morozov) [inline src/Types/Type.php:24]: This needs better alignment.

File: docs/en/reference/types.rst
Comment (morozov) [inline docs/en/reference/types.rst:593]: Should it be `ascii_string` as a data type?

File: docs/en/reference/types.rst
Comment (morozov) [inline docs/en/reference/types.rst:163]: Should this be “non-unicode”?

File: src/ParameterType.php
Comment (morozov) [inline src/ParameterType.php:45]: It's better to use “ASCII” here since it's an acronym.

**Review by morozov — COMMENTED** [review]
Looks good. Please address the remaining comments and squash. It should be good to go.

**Review by morozov — DISMISSED** [review]

**Comment by greg0ire** [comment]
In your example:

```php
$statement = $connection->prepare(
   'SELECT * FROM ... WHERE myVarcharColumn = :value' -- will bind as N'...'
);
```

What exactly dictates that `:value` will be bound as `N'…'`?

**Comment by gjdanis** [comment]
@greg0ire the driver code prepends the `N` character. If you use SQL Server Profiler you can see that the SQL that is sent to the server includes it. The actual code for this can be found here: https://github.com/microsoft/msphpsql/blob/5f76f838e5d44ca951a44bec6be1af05fae4bd93/source/pdo_sqlsrv/pdo_dbh.cpp#L1565

[Here's](https://github.com/microsoft/msphpsql/issues/1018) my conversation with the developers of the driver about this isssue. There's also an [RFC](https://wiki.php.net/rfc/extended-string-types-for-pdo) about this issue from the PDO maintainers. This issue is why they added the extended string types.

**Review by greg0ire — DISMISSED** [review]
Ok I think I got this right thanks to your explanation: when using the new ascii type, `PDO::SQLSRV_ENCODING_SYSTEM` will be used which means, "bind as 8-bit chars", which in turns means the driver will not add the `N` that would result in the implicit cast of the column values to utf-8

**Comment by greg0ire** [comment]
Should this target 2.11.x or is there a BC-break somewhere in there?

**Comment by gjdanis** [comment]
@greg0ire I think it could target 2.11.x but a previous conversation with @morozov asked that I move it to 3.0.x: https://github.com/doctrine/dbal/pull/4270#issuecomment-691362520

Let me know if I should retarget.

**Comment by greg0ire** [comment]
He was asking you to target to 3.0.x, but that retarget was from master… @morozov should we re-evaluate this? I see no reason to postpone this to 3.0.x when we could have it in 2.11.x .

File: src/Platforms/AbstractPlatform.php
Comment (greg0ire) [inline src/Platforms/AbstractPlatform.php:196]: Any reason not to use a type declartion here?

```suggestion
    public function getAsciiStringTypeDeclarationSQL(array $column): string
```

File: src/Platforms/SQLServer2012Platform.php
Comment (greg0ire) [inline src/Platforms/SQLServer2012Platform.php:1253]:
```suggestion
    public function getAsciiStringTypeDeclarationSQL(array $column): string
```

File: src/Types/AsciiStringType.php
Comment (greg0ire) [inline src/Types/AsciiStringType.php:10]: @morozov I see we never use `final` for types, but maybe we should start?

**Review by greg0ire — COMMENTED** [review]
A few extra comments while you wait for a decision on the target branch

File: src/Types/AsciiStringType.php
Comment (morozov) [reply src/Types/AsciiStringType.php:10]: Absolutely. The requirement for the return type declaration is enforced by the coding standard only in `4.x` but the use of final classes isn't enforced even there. It might be a good topic to explore further at the standard level.

File: src/Platforms/AbstractPlatform.php
Comment (morozov) [reply src/Platforms/AbstractPlatform.php:196]: > Any reason not to use a type declaration here?

I believe it's because all other methods don't have it and it's not enforced by the standard. Although it's a good thing to use them for new code.

**Comment by morozov** [comment]
> @morozov should we re-evaluate this? I see no reason to postpone this to 3.0.x when we could have it in 2.11.x .

I don't see a problem having it in `2.11.x` I asked for `3.0.x` because I didn't foresee it being ready before `2.11.0` is released.

@gjdanis if you want to retarget the PR, please do not open a new one. Edit and rebase this one instead.

**Review by greg0ire — DISMISSED** [review]
@gjdanis I addressed the changes for you, LGTM

**Comment by greg0ire** [comment]
Wait why did you change the branch back to 3.0.x?

**Comment by greg0ire** [comment]
Ok this is why… I'll do the rebase for you.

**Comment by gjdanis** [comment]
@greg0ire I got a lot of conflicts so I changed the base branch back. Was trying to clean this up locally but it's a little messy. If you know of a better way to retarget please go ahead and do so.

**Comment by greg0ire** [comment]
There are way more conflicts than I expected, the code looks nothing like 3.0.x… let's change the base branch again I guess :sweat_smile: … 3.0.0 is not that far away anyway :crossed_fingers:

**Review by greg0ire — DISMISSED** [review]
LGTM, you tell us if you want to give the retarget another attempt or if you can wait for 3.0.0

**Comment by morozov** [comment]
Note that if we retarget at `2.11.x` and it turns out to be a lot of work, we'll also have to do the inverse of that work again when merging the result up. Although I'm fine with this and I can help to retarget it as well.

**Comment by gjdanis** [comment]
@greg0ire, @morozov -- what do you think about merging this PR into 3.0.x and opening a new PR for 2.11.x? When we merge 2.11.x into 3.0.x we can just drop the PR that implements this change in 2.11.x, since the code will already be there.

Let me know what you want to do. I'm happy to implement the changes in 2.11.x but I think it would be a lot to do in this PR.

**Comment by morozov** [comment]
@gjdanis it would be at least the same amount of work or even more. If you want this to be released in `2.11.0`, we should retarget this PR against `2.11.x`. I can do this myself.

**Comment by gjdanis** [comment]
@morozov I feel like there should be 2.11.x version of this code so that large codebases that are reliant on 2.0 versions of this library can benefit from the change without having to refactor their code to be 3.0.x compliant. If you're offering to do this work, I'll take you up on the offer 👍

File: lib/Doctrine/DBAL/Driver/PDOSqlsrv/Statement.php
Comment (morozov) [inline lib/Doctrine/DBAL/Driver/PDOSqlsrv/Statement.php:30]: I changed `PDO::PARAM_STR` to `ParameterType::STRING` during the rebase since the direct usage of `PDO` binding types in calls to DBAL is deprecated (although may work since they use the same underlying vaules).

File: tests/Doctrine/Tests/DBAL/Functional/ParameterTypes/AsciiTest.php
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/ParameterTypes/AsciiTest.php:22]: This had to be reworked during the rebase because in `2.x` `Statement::execute()` doesn't return a result.

**Comment by morozov** [comment]
@gjdanis please see if the rebased PR looks do you.

File: lib/Doctrine/DBAL/Driver/PDOSqlsrv/Statement.php
Comment (gjdanis) [reply lib/Doctrine/DBAL/Driver/PDOSqlsrv/Statement.php:30]: Makes sense 👍

File: tests/Doctrine/Tests/DBAL/Functional/ParameterTypes/AsciiTest.php
Comment (gjdanis) [reply tests/Doctrine/Tests/DBAL/Functional/ParameterTypes/AsciiTest.php:22]: Looks good thanks!

**Review by gjdanis — COMMENTED** [review]
@morozov changes LGTM - thanks for the work to rebase!

**Review by morozov — APPROVED** [review]

**Comment by morozov** [comment]
Thanks, @gjdanis!

**Comment by greg0ire** [comment]
It sounds like a good plan to me!

**Comment by greg0ire** [comment]
Strategy for the merge-up:
1. merge 2.11.x as it was before merging this PR (so `git merge cce615a9e`)
2. merge this PR as it was before the rebase ( `git merge 3b8c8f9f6`)
3. merge 2.11.x as it is/was after merging this PR, but discard changes (`git merge -s ours 2f3d9bb61`)
4. if 2.11.x moved since then, merge it.

**Pull request merged**
